File: src/main/java/com/bitwise/dominos/basicingredients/Topping.java
package com.bitwise.dominos.basicingredients;

/**
 * Created by Aniruddha Sinha on 24-06-2016.
 */
public class Topping {
    private String name;
    private double price;
    public Topping(String n,double p){
        this.name=n;
        this.price=p;
        validateName();
        validatePrice();
    }

    private void validatePrice() {
        if(this.price<10.0) throw new Topping.InvalidToppingPriceException();
    }
    public void validateName(){
        if(this.name.isEmpty()) throw new Topping.InvalidToppingNameException();
    }



    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Topping topping = (Topping) o;

        if (Double.compare(topping.price, price) != 0) return false;
        return name.equals(topping.name);

    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        result = name.hashCode();
        temp = Double.doubleToLongBits(price);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        return result;
    }

    public double getPrice() {
        return price;
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "Topping{" +
                "name='" + name + '\'' +
                ", price=" + price +
                '}';
    }

    public class InvalidToppingPriceException extends RuntimeException {
    }

    public class InvalidToppingNameException extends RuntimeException {
    }
}

File: src/main/scala/aniruddha/dominos/basic_ingredients/Topping.scala
package aniruddha.dominos.basic_ingredients

/**
  * Created by Aniruddha Sinha on 26-06-2016.
  */
class Topping(n:String,p:Double) {
  val name=n
  val price=p
  def getName:String=name
  def getPrice:Double=price
  override def toString:String="Topping["+name+","+price+"]"

  def canEqual(other: Any): Boolean = other.isInstanceOf[Topping]

  override def equals(other: Any): Boolean = other match {
    case that: Topping =>
      (that canEqual this) &&
        name == that.name &&
        price == that.price
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(name, price)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}

object Topping{
  def apply(name:String,price:Double):Topping={
    if(name.isEmpty) throw new Topping.InvalidToppingNameException
    else if(price<10.0) throw new Topping.InvalidPriceValueException
    else new Topping(name,price)
  }


  case class InvalidToppingNameException() extends RuntimeException

  case class InvalidPriceValueException() extends RuntimeException
}